### ISSUE TITLE
Ambiguous Access Warning in ofGLRenderer / ofGLProgrammableRenderer

### ISSUE THREAD

**Issue by bakercp** [openframeworks/openFrameworks#3517]
@arturoc In OSX, I'm getting the following warnings in the latest master:

```
...openFrameworks/libs/openFrameworks/gl/ofGLRenderer.cpp:1784:29: Lookup of 'map' in member access expression is ambiguous; using member of 'ofBufferObject'
```
```
...openFrameworks/libs/openFrameworks/gl/ofGLProgrammableRenderer.cpp:2474:29: Lookup of 'map' in member access expression is ambiguous; using member of 'ofBufferObject'
```

**Comment by arturoc** [comment]
yes, i've seen that, seems like clang has some trouble differentiating 2 functions with the same name with and without template, it doesn't happen with gcc, it might also be something that is legal under c++11 and it won't be solved until we have support in osx. i'd wait till then and then decide, perhaps we can change the templated method to mapTyped

**Comment by kylemcdonald** [comment]
`map` seems like the right name. instead, discussion here suggests we disable the warning globally with `-Wambiguous-member-template`:

http://stackoverflow.com/questions/20326790/workaround-clang-warning-wambiguous-member-template

**Comment by bilderbuchi** [comment]
I gotta say, I like the "don't use `using namespace std`" [answer](http://stackoverflow.com/a/20327378/599884) more - globally disabling warnings to get away with this feels wrongish to me.

**Comment by bakercp** [comment]
I'm with @bilderbuchi on this one.  Start by beginning to remove using namespace std in the core and move those to the examples (as was discussed in the Roadmap ... somewhere).

**Comment by arturoc** [comment]
i agree that we should remove the global `using namespace std` perhaps by including it in ofMain.h instead of ofConstants.h at first. That way old projects, most addons... will still compile but people who want to opt out can include individual headers which surely some advanced users already do.

I'm not sure, though, that in this case this is the problem since the map seems to be conflicting with a different map function (not templated) in the same object not with anything global in std.

**Comment by bilderbuchi** [comment]
you don't think it's conflicting with `std::map` dragged in by the `using`?

**Comment by arturoc** [comment]
not 100% sure but i don't think so. since it's a method of a class, not a function, it's always called like `buffer.map()` or `buffer.map<type>()` which shouldn't clash with `std::map` i think

**Comment by admsyn** [comment]
Xcode does say the ambiguity is caused by `std::map`

An option is to wrap the line with clang pragmas like this:
```cpp
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wambiguous-member-template"
unsigned char * p = buffer.map<unsigned char>(GL_READ_ONLY);
#pragma clang diagnostic pop
```
..which could be done either around the line or just ofGLRenderer.cpp in general. IMO doing it around the file with an additional #ifdef to only hit Xcode is the least noisy option. I can PR this if that's the way to go.

I agree that the "real" fix is removing `using namespace std` but this may be a pragmatic (ha) fix for now to keep 0.9.0 progress in motion.

**Comment by bilderbuchi** [comment]
well, it's "just" a warning, and it's doing the right thing (I think) currently, so I think we could just leave it in for now (instead of clang-ignoring it), and implement a "proper" fix down the road, when 0.9.0 is out the door. Leaving it in will at least keep reminding us that we should fix it. :-)
